src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIPhysicalHandle.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.Data.SqlClient.SNI
+{
+    internal abstract class SNIPhysicalHandle : SNIHandle
+    {
+        protected const int DefaultPoolSize = 4;
+#if DEBUG
+        private static int s_packetId;
+#endif
+        private SNIPacketPool _pool;
+
+        protected SNIPhysicalHandle(int poolSize = DefaultPoolSize)
+        {
+            _pool = new SNIPacketPool(poolSize);
+        }
+
+        public override SNIPacket RentPacket(int headerSize, int dataSize)
+        {
+            SNIPacket packet;
+            if (!_pool.TryGet(out packet))
+            {
+#if DEBUG
+                int id = Interlocked.Increment(ref s_packetId);
+                packet = new SNIPacket(this, id);
+#else
+                packet = new SNIPacket();
```

Comment:
Can you also write eventsource trace logs for various events with packet Id to trace SNI packet activity?

Reply:
That would only work with Debug builds which would have to be supplied to the user directly not through Nuget. I can probably add trace events for the more common things like renting returning and possible rent cache misses in release builds if you want.

Reply:
Something that can help trace packet re-use would be beneficial I think.

Reply:
Currently in Debug builds there's a packet id which is allocated from a global interlocked increment so each packet has a unique id for the lifetime of the process. I can make that id available in release mode. If that id is available i can event source events into rent and return on the physical handle. 
Unfortunately that just means that when you're debugging a lost packet you see the ones that are behaving correctly but you'll never see the incorrect ones until they hit GC and to see them at that point we'd have to have a trace call in the finalizer thread writing output which doesn't sound like a good idea.

As you can see there's a lot of debug only code in the pysical packet which deals with the packet lifetimes. Enabling that functionality makes everything very very slow and theres so much information that you need to deal with that it's only useful when you've got complete knowledge of individual packets in a simple problem case.

I agree with the principle of being able to trace the packets but i'm not sure how to usefully trace much beyond pool rental (pool hit and miss) and return without significant overhead. Let me know what you think.

Reply:
Fair enough, I don't think it's super important for end users to be able to capture that info. I agree debugging with driver will give us the information we need.